File: python/packages/core/src/matimo/core/tfidf_embedding.py
"""
TF-IDF Embedding Provider — zero-dependency semantic search.

Mirrors: packages/core/src/core/tfidf-embedding.ts

Lightweight TF-IDF (Term Frequency–Inverse Document Frequency) for
cosine-similarity ranking. Good enough for 10–200 skills. For production
enterprise deployments, plug in an OpenAI/Cohere embedding provider via the
``EmbeddingProvider`` protocol instead.
"""
from __future__ import annotations

import math
import re
from typing import Protocol, runtime_checkable

# ---------------------------------------------------------------------------
# EmbeddingProvider protocol (mirrors TS EmbeddingProvider interface)
# ---------------------------------------------------------------------------


@runtime_checkable
class EmbeddingProvider(Protocol):
    """Pluggable embedding provider for semantic skill search."""

    @property
    def dimensions(self) -> int: ...

    async def embed(self, text: str) -> list[float]: ...

    async def embed_batch(self, texts: list[str]) -> list[list[float]]: ...


# ---------------------------------------------------------------------------
# TF-IDF provider
# ---------------------------------------------------------------------------

_STOPWORDS = frozenset(
    "a an the and or but in on at to for of with by from is it as be was are "
    "this that not do if so no up my we he she they you me us all can had has "
    "have will would could should may might shall been being were did does its "
    "than then also very just about more some any each every how what when where".split()
)

_SPLIT_RE = re.compile(r"[^a-z0-9]+")


class TfIdfEmbeddingProvider:
    """
    Simple TF-IDF based embedding provider.

    Builds a vocabulary from a registered corpus and represents each text as a
    TF-IDF weighted vector with L2 normalisation.
    """

    def __init__(self) -> None:
        self._vocabulary: dict[str, int] = {}
        self._idf: list[float] = []
        self._corpus_size: int = 0
        self._dimensions: int = 0

    @property
    def dimensions(self) -> int:
        return self._dimensions

    # ------------------------------------------------------------------
    # Fit vocabulary
    # ------------------------------------------------------------------

    def fit(self, documents: list[str]) -> None:
        """Build vocabulary and IDF weights from a corpus. Must be called before ``embed``."""
        self._vocabulary.clear()
        self._corpus_size = len(documents)

        doc_frequency: dict[str, int] = {}
        for doc in documents:
            unique_terms = set(self._tokenize(doc))
            for term in unique_terms:
                doc_frequency[term] = doc_frequency.get(term, 0) + 1

        idx = 0
        idf_values: list[float] = []
        for term, df in doc_frequency.items():
            self._vocabulary[term] = idx
            idx += 1
            # Smooth IDF: log((N+1) / (df+1)) + 1
            idf_values.append(math.log((self._corpus_size + 1) / (df + 1)) + 1)

        self._dimensions = len(idf_values)
        self._idf = idf_values

    # ------------------------------------------------------------------
    # Embed
    # ------------------------------------------------------------------

    async def embed(self, text: str) -> list[float]:
        return self.embed_sync(text)

    async def embed_batch(self, texts: list[str]) -> list[list[float]]:
        return [self.embed_sync(t) for t in texts]

    def embed_sync(self, text: str) -> list[float]:
        """Synchronous embed — no async overhead."""
        if self._dimensions == 0:
            return []

        terms = self._tokenize(text)
        tf = [0.0] * self._dimensions

        for term in terms:
            idx = self._vocabulary.get(term)
            if idx is not None:
                tf[idx] += 1.0

        # Sublinear TF: 1 + log(tf) if tf > 0, weighted by IDF
        for i in range(self._dimensions):
            if tf[i] > 0:
                tf[i] = (1 + math.log(tf[i])) * self._idf[i]

        # L2 normalise
        norm = math.sqrt(sum(v * v for v in tf))
        if norm > 0:
            tf = [v / norm for v in tf]

        return tf

    # ------------------------------------------------------------------
    # Tokenise
    # ------------------------------------------------------------------

    @staticmethod
    def _tokenize(text: str) -> list[str]:
        return [
            t
            for t in _SPLIT_RE.split(text.lower())
            if len(t) > 1 and t not in _STOPWORDS
        ]


# ---------------------------------------------------------------------------
# Cosine similarity
# ---------------------------------------------------------------------------


def cosine_similarity(a: list[float], b: list[float]) -> float:
    """Cosine similarity between two vectors (1 = identical, 0 = orthogonal)."""
    if len(a) != len(b) or not a:
        return 0.0

    dot = sum(ai * bi for ai, bi in zip(a, b, strict=False))
    norm_a = math.sqrt(sum(ai * ai for ai in a))
    norm_b = math.sqrt(sum(bi * bi for bi in b))

    denom = norm_a * norm_b
    return dot / denom if denom > 0 else 0.0

File: typescript/packages/core/src/core/tfidf-embedding.ts
/**
 * TF-IDF Embedding Provider — zero-dependency semantic search
 * (Term Frequency–Inverse Document Frequency)
 * Provides a lightweight text-to-vector implementation using TF-IDF (Term
 * Frequency–Inverse Document Frequency) for cosine-similarity ranking. This
 * is good enough for 10–200 skills. For production enterprise deployments,
 * plug in an OpenAI/Cohere `EmbeddingProvider` instead.
 *
 * No external dependencies — works out of the box.
 */

import { EmbeddingProvider } from './types.js';

/**
 * Simple TF-IDF based embedding provider.
 * Builds a vocabulary from the registered corpus and represents each text as
 * a TF-IDF weighted vector.
 */
export class TfIdfEmbeddingProvider implements EmbeddingProvider {
  private vocabulary: Map<string, number> = new Map();
  private idf: Float64Array = new Float64Array(0);
  private corpusSize = 0;
  private _dimensions = 0;

  get dimensions(): number {
    return this._dimensions;
  }

  /**
   * Build the vocabulary and IDF weights from a corpus of documents.
   * Must be called before `embed()` or `embedBatch()`.
   */
  fit(documents: string[]): void {
    this.vocabulary.clear();
    this.corpusSize = documents.length;

    // Build vocabulary (unique terms across all documents)
    const docFrequency = new Map<string, number>();
    for (const doc of documents) {
      const terms = this.tokenize(doc);
      const uniqueTerms = new Set(terms);
      for (const term of uniqueTerms) {
        docFrequency.set(term, (docFrequency.get(term) || 0) + 1);
      }
    }

    // Assign indices and compute IDF
    let idx = 0;
    const idfValues: number[] = [];
    for (const [term, df] of docFrequency) {
      this.vocabulary.set(term, idx++);
      // Smooth IDF: log((N + 1) / (df + 1)) + 1
      idfValues.push(Math.log((this.corpusSize + 1) / (df + 1)) + 1);
    }

    this._dimensions = idfValues.length;
    this.idf = new Float64Array(idfValues);
  }

  async embed(text: string): Promise<number[]> {
    return this.embedSync(text);
  }

  async embedBatch(texts: string[]): Promise<number[][]> {
    return texts.map((t) => this.embedSync(t));
  }

  /**
   * Synchronous embed for internal use (no async overhead).
   */
  embedSync(text: string): number[] {
    if (this._dimensions === 0) {
      return [];
    }

    const terms = this.tokenize(text);
    const tf = new Float64Array(this._dimensions);

    // Count term frequencies
    for (const term of terms) {
      const idx = this.vocabulary.get(term);
      if (idx !== undefined) {
        tf[idx]++;
      }
    }

    // Normalize TF (sublinear: 1 + log(tf) if tf > 0)
    for (let i = 0; i < this._dimensions; i++) {
      if (tf[i] > 0) {
        tf[i] = (1 + Math.log(tf[i])) * this.idf[i];
      }
    }

    // L2 normalize
    let norm = 0;
    for (let i = 0; i < this._dimensions; i++) {
      norm += tf[i] * tf[i];
    }
    norm = Math.sqrt(norm);
    if (norm > 0) {
      for (let i = 0; i < this._dimensions; i++) {
        tf[i] /= norm;
      }
    }

    return Array.from(tf);
  }

  /**
   * Tokenize text into lowercase terms.
   * Splits on non-alphanumeric characters and filters stopwords.
   */
  private tokenize(text: string): string[] {
    return text
      .toLowerCase()
      .split(/[^a-z0-9]+/)
      .filter((t) => t.length > 1 && !STOPWORDS.has(t));
  }
}

/**
 * Cosine similarity between two vectors.
 * Returns a value between -1 and 1 (1 = identical, 0 = orthogonal).
 */
export function cosineSimilarity(a: number[], b: number[]): number {
  if (a.length !== b.length || a.length === 0) return 0;

  let dot = 0;
  let normA = 0;
  let normB = 0;

  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }

  const denominator = Math.sqrt(normA) * Math.sqrt(normB);
  return denominator === 0 ? 0 : dot / denominator;
}

/** Common English stopwords to exclude from TF-IDF */
const STOPWORDS = new Set([
  'a',
  'an',
  'the',
  'and',
  'or',
  'but',
  'in',
  'on',
  'at',
  'to',
  'for',
  'of',
  'with',
  'by',
  'from',
  'is',
  'it',
  'as',
  'be',
  'was',
  'are',
  'this',
  'that',
  'not',
  'do',
  'if',
  'so',
  'no',
  'up',
  'my',
  'we',
  'he',
  'she',
  'they',
  'you',
  'me',
  'us',
  'all',
  'can',
  'had',
  'has',
  'have',
  'will',
  'would',
  'could',
  'should',
  'may',
  'might',
  'shall',
  'been',
  'being',
  'were',
  'did',
  'does',
  'its',
  'than',
  'then',
  'when',
  'what',
  'which',
  'who',
  'how',
  'there',
  'here',
  'about',
  'into',
  'over',
  'after',
  'also',
  'each',
  'just',
  'only',
  'very',
]);
